src/utils/bounded_priority_queue.rs

```diff
@@ -0,0 +1,119 @@
+use crate::serializable_traits::Data;
+use serde_derive::{Deserialize, Serialize};
+use std::cmp::Reverse;
+use std::collections::BinaryHeap;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub(crate) struct BoundedMinPriorityQueue<T: Ord> {
+    max_size: usize,
+    underlying: BinaryHeap<Reverse<T>>,
+}
+
+impl<T: Data + Ord> BoundedMinPriorityQueue<T> {
+    pub fn new(max_size: usize) -> BoundedMinPriorityQueue<T> {
+        BoundedMinPriorityQueue {
+            max_size: max_size,
+            underlying: BinaryHeap::with_capacity(max_size),
+        }
+    }
+
+    pub fn get_size(&self) -> usize {
+        self.underlying.len()
+    }
+
+    pub fn into_vec_sorted(&self) -> Vec<T> {
```

Comment:
To be idiomatic instead we would implement `Into<Vec<T>>` trait for `BoundedMinPriorityQueue<T>` and then call `.into()` where necessary (or even if we have to be generic over the receiving type on the call site where this is necesssary with something like `where C: Into<Vec<T>>, whatever makes more sense).